### PR TITLE
Dependency Injection

### PR DESCRIPTION
**Pull request by Vadifire**
In addition to the Service->Resource dependency injection, I also used Dagger 2 to inject Twitter4J's Twitter API into my Twitter4JService class. I'm not sure if this was required, but it seemed in spirit of the DI goals defined here: [https://en.wikipedia.org/wiki/Dependency_injection#Intent](url) as it further decouples configuration.

### CODE REVIEW

File: src/main/java/tweeting/services/Twitter4JService.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        } else {
+            return new TwitterServiceResponseException(te);
+        }
+    }
+
+    private Tweet constructTweet(Status status) {
+        if (status == null) {
+            return null;
+        }
+        Tweet tweet = new Tweet();
```
Comment (andrewcchau) [inline src/main/java/tweeting/services/Twitter4JService.java]: final

File: src/main/java/tweeting/services/Twitter4JService.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(Twitter4JService.class);
+
+    @Inject
+    public Twitter4JService(Twitter api) {
+        this.api = api;
+    }
+
+    @Override
+    public Optional<Tweet> postTweet(String message) throws TwitterServiceResponseException,
+            TwitterServiceCallException {
```
Comment (andrewcchau) [inline src/main/java/tweeting/services/Twitter4JService.java]: Keep exceptions on the same line

File: src/main/java/tweeting/services/Twitter4JService.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                        return constructTweetList(statuses);
+                    });
+        } catch (TwitterException te) {
+            throw createServerException(te);
+        }
+    }
+
+    @Override
+    public Optional<List<Tweet>> getFilteredTimeline(String keyword) throws TwitterServiceResponseException,
+            TwitterServiceCallException {
```
Comment (andrewcchau) [inline src/main/java/tweeting/services/Twitter4JService.java]: keep exceptions on the same line

File: src/main/java/tweeting/services/Twitter4JService.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    private Tweet constructTweet(Status status) {
+        if (status == null) {
+            return null;
+        }
+        Tweet tweet = new Tweet();
+        tweet.setMessage(status.getText());
+        if (status.getUser() == null) {
+            logger.warn("Tweet has no user.");
+        } else {
+            TwitterUser user = new TwitterUser();
```
Comment (andrewcchau) [inline src/main/java/tweeting/services/Twitter4JService.java]: final

File: src/test/java/tweeting/resources/TwitterResourceTest.java [hunk 5]
```diff
@@ -22,7 +23,7 @@
 
 public class TwitterResourceTest {
 
-    // Mocked classes
+    // Mocked service
     TwitterService service;
```
Comment (andrewcchau) [inline src/test/java/tweeting/resources/TwitterResourceTest.java]: private

File: src/main/java/tweeting/injection/modules/TweetingModule.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Provides
+    @Singleton
+    static LogFilter provideLogFilter() {
+        return new LogFilter();
+    }
+
+    @Provides
+    @Singleton
+    static TwitterOAuthCredentials provideTwitterAuthentication(TweetingConfiguration conf) {
```
Comment (kev-chan) [inline src/main/java/tweeting/injection/modules/TweetingModule.java]: why the need for this and provideTwitterAPI in the other module class?

File: src/test/java/tweeting/services/TwitterServiceTest.java [hunk 7]
```diff
@@ -28,34 +27,33 @@
 public class TwitterServiceTest {
```
Comment (kev-chan) [inline src/test/java/tweeting/services/TwitterServiceTest.java]: why is this class called TwitterServiceTest?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/tweeting/injection/modules/Twitter4JModule.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ * Note: TwitterOAuthCredentials are required for configuring the API the service relies on.
+ */
+
+@Module
+public class Twitter4JModule {
+
+    @Provides
+    @Singleton
+    static TwitterService provideTwitterService(TweetingConfiguration conf) {
+        final TwitterOAuthCredentials auth = conf.getTwitterAuthorization();
```
Comment (kev-chan) [inline src/main/java/tweeting/injection/modules/Twitter4JModule.java]: add null check, also, this likely does not need to be static method

**Pull request merged**